### PR TITLE
Extensions to the ObjectRegistry 

### PR DESCRIPTION
**Pull request by mrambacher**
The ObjectRegistry is now a member variable of the DBOptions and ConfigOptions.  This allows a registry to be on a per-configuration basis.

The ObjectLibrary was expanded to include a "Program" and a "Loaded" library subclass.  The "Loaded" library is initialized from a shared library.  The "Program" library is initialized from an in-process method.

### CODE REVIEW

File: include/rocksdb/utilities/object_registry.h
Comment (zhichao-cao) [inline include/rocksdb/utilities/object_registry.h:165]: What's the difference between local library and dynamic library? May need some explanations and examples.

File: include/rocksdb/utilities/object_registry.h
Comment (zhichao-cao) [inline include/rocksdb/utilities/object_registry.h:83]: How can a user decide whether a local, or dynamic should be used?

File: include/rocksdb/utilities/object_registry.h
Comment (zhichao-cao) [inline include/rocksdb/utilities/object_registry.h:68]: Dump is in the public API, if some one is using it, how they can write out the info in registry to the log?

File: include/rocksdb/utilities/object_registry.h
Comment (zhichao-cao) [inline include/rocksdb/utilities/object_registry.h:100]: It is a little confusion to use "CreateFromString", do you have other better name? Or you want to align it with the "String" in Configurable?

File: include/rocksdb/utilities/object_registry.h
Comment (zhichao-cao) [inline include/rocksdb/utilities/object_registry.h:113]: If the API is not used in the current stage, do you think we can add it later to the code base?

File: include/rocksdb/utilities/object_registry.h
Comment (mrambacher) [reply include/rocksdb/utilities/object_registry.h:165]: I will add a comment later.

A dynamic library is one that was loaded into the application explicitly.  Methods are located by name in that external library.

A "local library" searches the current executable for the method in question, if enabled/allowed.  Methods in a "local library" could be searched by name (if allowed) or their address might be known.  In the latter case, the local library could not be reconstructed on restart.

If you have a better name for the "local library", feel free to suggest one.

File: include/rocksdb/utilities/object_registry.h
Comment (mrambacher) [reply include/rocksdb/utilities/object_registry.h:83]: If you are loading an external method, it will be a dynamic library.  If you are loading a method from the current executable, it could be put into the local or default library.

File: include/rocksdb/utilities/object_registry.h
Comment (mrambacher) [reply include/rocksdb/utilities/object_registry.h:100]: All Customizable classes will have a CreateFromString method.  This is the "standard" constructor method used by the options hierarchy

File: include/rocksdb/utilities/object_registry.h
Comment (mrambacher) [reply include/rocksdb/utilities/object_registry.h:113]: It could be added later, but is useful if we want to put back Dump...

File: include/rocksdb/utilities/object_registry.h
Comment (zhichao-cao) [reply include/rocksdb/utilities/object_registry.h:100]: Oh, yeah, I have the impression that CreateFromString is the member function.

File: include/rocksdb/utilities/object_registry.h
Comment (zhichao-cao) [reply include/rocksdb/utilities/object_registry.h:83]: I'm thinking another name to replace "Local". "local" is not an exclusion case of "dynamic"......

File: include/rocksdb/utilities/object_registry.h
Comment (zhichao-cao) [inline include/rocksdb/utilities/object_registry.h:117]: Is there any order of the names being returned? Maybe just explain it in comment is fine.

File: include/rocksdb/utilities/object_registry.h
Comment (zhichao-cao) [inline include/rocksdb/utilities/object_registry.h:169]: I'm little confused by the two "AddLocalLibrary" and two "AddDynamicLibrary". Can you explain more?

File: include/rocksdb/utilities/object_registry.h
Comment (zhichao-cao) [inline include/rocksdb/utilities/object_registry.h:300]: It is confusion that the function name is "GetRegisteredNames" but returns the number of names of certain type.

File: include/rocksdb/utilities/object_registry.h
Comment (mrambacher) [reply include/rocksdb/utilities/object_registry.h:169]: For local, the one signature takes in a name (the method argument) of a registrar function and attempts to find that function in the executable using the Env DynamicLibrary methods.  If the function was not found, an error is returned.  

The other method takes in the RegistrarFunction (a std::function) and invokes it to register the types in the library.  The first method will end up calling this one if the function is found.

Comparable signatures exist for the dynamic libraries, using a DynamicLibrary to locate the registrar function and the Env to locate the dynamic library.

The "string-based" signature must be used in order to recover/restore the ObjectRegistry from a string (or options file).  The "registrar" signature is what does the actual registration, and is useful for programs who wish to register their types at "compile-time".

File: include/rocksdb/utilities/object_registry.h
Comment (mrambacher) [reply include/rocksdb/utilities/object_registry.h:300]: Would "GetNamesForType" be better?

File: utilities/object_registry.cc
Comment (zhichao-cao) [inline utilities/object_registry.cc:217]: Make it clear that, the returned value is the sum of all factories registered for all types, "num_types" only contains the types

File: utilities/object_registry.cc
Comment (zhichao-cao) [inline utilities/object_registry.cc:253]: Add some simple comments here for reference.

File: env/env.cc
Comment (zhichao-cao) [inline env/env.cc:84]: In HISTORY.md we need to mention the return code changes here.

File: env/env_posix.cc
Comment (zhichao-cao) [inline env/env_posix.cc:113]: This change is not related to current OBjectRegistry right?

File: include/rocksdb/utilities/object_registry.h
Comment (zhichao-cao) [reply include/rocksdb/utilities/object_registry.h:169]: Thanks for the explanation. Maybe you can add those to the comments of "AddLocalLibrary" and "AddDynamicLibrary"

File: include/rocksdb/utilities/object_registry.h
Comment (zhichao-cao) [reply include/rocksdb/utilities/object_registry.h:300]: Yes, let's use "GetNamesForType"

File: include/rocksdb/utilities/object_registry.h
Comment (zhichao-cao) [inline include/rocksdb/utilities/object_registry.h:307]: How about "GetFactoryNums" for this one?

File: options/db_options.cc
Comment (zhichao-cao) [inline options/db_options.cc:487]: I'm not very clear about the logic here. If `copy.registry == immutable_.object_registry` why we need 'immutable_.object_registry = copy.registry->Clone()'.

File: utilities/object_registry.cc
Comment (zhichao-cao) [inline utilities/object_registry.cc:148]: I'm not sure `Clone` is a good function name here. We just share the same OBjectRegistry instance pointer right? Not clone a new instance.

File: utilities/object_registry.cc
Comment (zhichao-cao) [inline utilities/object_registry.cc:238]: In comment "looking for the an entry" no "the"?

File: utilities/object_registry.cc
Comment (zhichao-cao) [inline utilities/object_registry.cc:261]: I several places, the variable name "copy" and "left" are used. Maybe replacing with more explicit name.

File: include/rocksdb/utilities/object_registry.h
Comment (zhichao-cao) [inline include/rocksdb/utilities/object_registry.h:43]: Add some comments to explain 1) what is ObejctLibrary, 2) why it is inherited from Customizable.

**Review by zhichao-cao — APPROVED** [review]
LGTM, CC @ajkr if he has further comments?

File: include/rocksdb/utilities/object_registry.h
Comment (ajkr) [inline include/rocksdb/utilities/object_registry.h:252]: Is there a way to verify the plugin was built against compatible headers?
Also I wonder if there is any effort to expand our ABI compatibility guarantees -- that should make this approach more compelling. Currently we only do best-effort to uphold ABI compatibility across patch versions within a minor version.

**Review by ajkr — COMMENTED** [review]
Is the title up to date? Somehow I couldn't find any of the things mentioned in the title.

edit: Upon further thought I think we should prioritize #6614 over this by a lot. 6614 has general benefit to extension users/implementors such as FB, whereas this PR has less general benefits. I know you already submitted 6614 a year ago, so we just need to review our review priorities...

File: include/rocksdb/utilities/object_registry.h
Comment (mrambacher) [reply include/rocksdb/utilities/object_registry.h]: This is not part of the current submission.  The PR was broken into more pieces and this will be in a follow-on

**Comment by mrambacher** [comment]
> Is the title up to date? Somehow I couldn't find any of the things mentioned in the title.
> 
> edit: Upon further thought I think we should prioritize #6614 over this by a lot. 6614 has general benefit to extension users/implementors such as FB, whereas this PR has less general benefits. I know you already submitted 6614 a year ago, so we just need to review our review priorities...

The title and comment were out of date.  The scope of the PR was narrowed after the "plugin" work came into existence.

I plan to update #6614 shortly and submit it as a bunch of separate PRs (few Customizable objects per class).

File: include/rocksdb/utilities/object_registry.h
Comment (mrambacher) [reply include/rocksdb/utilities/object_registry.h:252]: In #7949 I propose how to do better ABI validation.  I can remove this method until that is approved if that helps this one move forward.

Comment (mrambacher) [reply include/rocksdb/utilities/object_registry.h]: I removed the AddLoadedLibrary method for now.

File: options/db_options.cc
Comment (mrambacher) [reply options/db_options.cc]: Code restructured.  No longer needed

File: env/env_posix.cc
Comment (mrambacher) [reply env/env_posix.cc]: Removed from this PR

**Review by ajkr — CHANGES_REQUESTED** [review]
The title/description appear to not reflect the contents of the PR. They should contain a convincing reason to add complexity to the public API.

edit: I was not able to figure out the use case from reading the code. If it is not strictly needed, it can save reviewer and maintenance effort to not pursue it.

**Comment by riversand963** [comment]
Haven't reviewed the entire series of PRs about configuration and object registry. Read about dlopen/dlsym a bit more..
If dlsym is used to access symbols defined in the shared library, then it seems that we cannot enforce function signature check on functions loaded from the shared library.

Will this be an issue if RocksDB changes the API in a future release?  Can we have some compile/link time check?

**Comment by riversand963** [comment]
Something else not related to this PR.
The use of `std::regex` (https://github.com/facebook/rocksdb/blob/master/include/rocksdb/utilities/object_registry.h#L61) is treated as an error by our internal linter due to its recursion on program stack, https://gcc.gnu.org/bugzilla/show_bug.cgi?id=61582, and it should be removed from code that can potentially be used in production.

**Comment by mrambacher** [comment]
Abandoning this PR now in favor of other approaches.

**Pull request closed without merging**
